[BREAKING] Multicolumn transformations for GoupedDataFrame (#2481)
JuliaData/DataFrames.jl: docs/src/man/split_apply_combine.md

> # Transforming data frames

Reviewer: This is probably too general. Better keep something more specific to grouping, at least for now, and maybe later reorganise docs so that we present the `select`/`transform`/`combine` API elsewhere in detail outside of the context of grouping.

Author: OK - I have reverted.

object from your data frame using the `groupby` function that takes two arguments:
(1) a data frame to be grouped, and (2) a set of columns to group by.

!!! note

    All operations described for `GroupedDataFrame` in this section of the manual
    are also supported for `AbstractDataFrame` in which case it is considered as
    being grouped on no columns (meaning it has a single group, or zero groups if it is empty).
    The only difference is that in this case the `keepkeys` and `ungroup` keyword

>     arguments are not supported and a data frame is always returned.

Reviewer: Not to be addressed by this PR, but it's weird that `select` and `transform` are documented in detail only in this section, which is about grouped operations.

Author: Agreed, but the point is that `select`, `transform`, and `combine` are now transparently handling data frames and grouped data frames in the same way - so maybe we just need to change the name of this section? As this note is about the case when "split" step is skipped.

Author: I have just renamed this section of the manual.

All these functions take a specification of one or more functions to apply to
each subset of the `DataFrame`. This specification can be of the following forms:
1. standard column selectors (integers, `Symbol`s, vectors of integers, vectors of symbols,
   `All`, `:`, `Between`, `Not` and regular expressions)
2. a `cols => function => target_cols` form additionally specifying the target column or columns
3. a `cols => function` pair indicating that `function` should be called with
   positional arguments holding columns `cols`, which can be a any valid column selector;
   in this case target column name is automatically generated and it is assumed that
   `function` returns a single value or a vector; the generated name is created by concatenating

>    source column name and `function` name where possible (see examples below).

Author: fixed

> 4. a `col => target_cols` pair, which renames the column `col` to `target_cols`

Author: fixed (but I keep the name, as we reference it below)

5. a `nrow` or `nrow => target_cols` form which efficiently computes the number of rows
   in a group (without `target_cols` the new column is called `:nrow`)

> 6. vectors or matrices transformations specified by `Pair` syntax described in points 2 to 5

Author: We allow vectors of symbols etc. (this is covered in point 1.)

The description of vararg is rewritten and now is below (https://github.com/JuliaData/DataFrames.jl/pull/2481/files#diff-45f87ad2a77d5bb7ac1969a14bb11079c730996f6904a82f9ae00222d1ae5f72R56)

8. a function which will be called with a `SubDataFrame` corresponding to each group;
   this form should be avoided due to its poor performance unless a very large
   number of columns are processed (in which case `SubDataFrame` avoids excessive
   compilation)

All functions have two types of signatures. One of them takes a `GroupedDataFrame`
as a first argument and an arbitrary number of transfomations described above
as following arguments. The second type of signature is when `Function` or `Type`
is passed as a first argument and `GroupedDataFrame` is a second argument (in a
similar fashion like it is passed in e.g. `map` function).

> As a special rule that applies to `cols => function` and `cols => function =>
> target_cols` syntaxes is the following. If `cols` is wrapped in an `AsTable`

Reviewer: ```suggestion
As a special rule, with the `cols => function` and `cols => function =>
target_cols` syntaxes, if `cols` is wrapped in an `AsTable`
```

object then a `NamedTuple` containing columns selected by `cols` is passed to
`function`.

> What is allowed for `function` to return is determined by the `target_cols` value
> in the following way:

Reviewer: ```suggestion
What is allowed for `function` to return is determined by the `target_cols` value:
```

> 1. If just a `function` is passed as an argument then returning a data frame,

Reviewer: ```suggestion
1. If `target_cols` is omitted, then returning a data frame,
```

Author: This is not correct. I mean specifically "a `function`" (so neither `cols` nor `target_cols`). If `cols` is passed but `target_cols` is omitted then we assume that **a single column is returned** (and we automatically generate a name for it). If you specify `cols` and want multiple columns you must be explicit (e.g. by adding `AsTable`) and in this case different rules apply.

   are set to be equal to the key names except if `keys` returns integers, in
   which case they are prefixed by `x` (so the column names are e.g. `x1`,
   `x2`, ...). If `target_cols` is a vector of `Symbol`s or strings then
   column names produced using the rules above are ignored and replaced by
   `target_cols` (the number of columns must be the same as the length of
   `target_cols` in this case).
   If `fun` returns a value of any other type then it is assumed that it is a
   table conforming to the Tables.jl API and the `Tables.columntable` function
   is called on it to get the resulting columns and their names. The names are
   retained when `target_cols` is `AsTable` and are replaced if
   `target_cols` is a vector of `Symbol`s or strings.

In all of these cases, `function` can return either a single row or multiple
rows. As a particular rule, values wrapped in a `Ref` or a `0`-dimensional
`AbstractArray` are unwrapped and then treated as a single row.

`select`/`select!` and `transform`/`transform!` always return a `DataFrame`
with the same number and order of rows as the source (even if `GroupedDataFrame`
had its groups reordered).

> For `combine` return value is ordered by the order of groups in `GroupedDataFrame`
> and for each group the functions can return an arbibrary number of rows (provided
> that these numbers are consistent).

Reviewer: ```suggestion
For `combine`, rows in the returned object appear in the order of
groups in the `GroupedDataFrame`. The functions can return an arbitrary number
of rows for each group, but the kind of returned object and the number
and names of columns must be the same for all groups.
```

It is allowed to mix single values and vectors if multiple transformations
are requested. In this case single value will be repeated to match the length
of columns specified by returned vectors.

To apply `function` to each row instead of whole columns, it can be wrapped in a
`ByRow` struct. In this case if `cols` is a `Symbol`, a string, or an
integer then `function` is applied to each element (row) of `cols` using
broadcasting. Otherwise `cols` can be any column indexing syntax, in
which case `function` will be passed one argument for each of the columns

> specified by `cols`. If `ByRow` is used it is allowed for

Reviewer: Do we really need to distinguish the 1-column case from the general case? This was needed back when a named tuple was passed, but now both cases follow the same rule.

Author: right. fixed

`cols` to select an empty set of columns, in which case `function`
 is called for each row without any arguments.

> The kind of return value and the number and names of columns must be the same for all groups.
> 

Reviewer: Remove if you accept my suggestion to mention it above.
```suggestion
```

There the following keyword arguments are supported by the transformation functions
(not all keyword arguments are supported in all cases; in general they are allowed
in situations when they are meaningful, see the documentation of the specific functions
for details):

> - `keepkeys` : if grouping columns should be kept in the returned data frame.
> - `ungroup` : if the retun value of the operation should be a data frame or a
>   `GroupedDataFrame`.
> - `copycols` : if columns of the source data frame should be copied if no transformation
>   is applied to them.
> - `renamecols` : if in `cols => funcion` form the automatically generated column name
>   should include the name of transformation function or not.

Reviewer: ```suggestion
- `keepkeys` : whether grouping columns should be kept in the returned data frame.
- `ungroup` : whether the return value of the operation should be a data frame or a
  `GroupedDataFrame`.
- `copycols` : whether columns of the source data frame should be copied if no transformation
  is applied to them.
- `renamecols` : whether in the `cols => function` form automatically generated column names
  should include the name of transformation functions or not.
```

│ 2   │ Iris-versicolor │ 50    │ 4.26    │
│ 3   │ Iris-virginica  │ 50    │ 5.552   │

julia> combine(gdf, [:PetalLength, :SepalLength] => ((p, s) -> (a=mean(p)/mean(s), b=sum(p))) =>

>        AsTable) # multiple columns are passed as arguments

Author: fixed
